Ui

khatamoto-cq opened the pull request:
UIの実装が完了

khatamoto-cq committed:
ログイン画面を作成

khatamoto-cq committed:
ログイン画面のテキストフィールドとログインボタンを調整

・テキストフィールドが見えにくいので枠をつけた
・テキストフィールドとボタンを角丸にした

khatamoto-cq committed:
ログインコントローラのクラス名をデフォルトのViewControllerからLoginViewControllerに変更

khatamoto-cq committed:
ログイン画面から書籍一覧の遷移を実装

khatamoto-cq committed:
バリデーションを追加

khatamoto-cq committed:
リファクタリング: UIControlの初期化を別のメソッドに切り出し

khatamoto-cq committed:
Tabを作成

khatamoto-cq committed:
アカウント設定画面を作成

khatamoto-cq committed:
ストーリーボードを修正

khatamoto-cq committed:
書籍一覧をViewControllerからTableViewControllerに変更

khatamoto-cq committed:
書籍一覧を実装

khatamoto-cq committed:
手間なのでメルアド、パスワードのチェックを外した

khatamoto-cq committed:
書籍編集のUIを作成

khatamoto-cq committed:
SwiftLintエラーを修正

khatamoto-cq committed:
warning: Unsupported Configuration: Plain Style unsupported in a Navigation Itemが表示されるのでUIBaserButtonItemのStyleを変更

khatamoto-cq committed:
書籍一覧に追加ボタン追加

khatamoto-cq committed:
追加画面のレイアウトを整えて、保存ボタンにアクションを付けた

khatamoto-cq committed:
不要なコメントを削除

khatamoto-cq committed:
Datepickerをテキストフィールドからポップアップするように修正

khatamoto-cq committed:
書籍追加画面で画像をアルバムから選択できるようにした。

khatamoto-cq committed:
書籍編集画面も同様にアルバムから画像を選択できるように修正

khatamoto-cq committed:
リファクタリング

・スマホではclickよりtapの方がアクションとしては最適なので修正
・IBOutletのサフィックスがオブジェクトのクラス名になっていない箇所を全体に合わせた
・書籍一覧のセル内の日付が購入日ではなく出版日になっていたので修正
・IBActionのサフィックスがActionになっていない箇所を全体に合わせた

khatamoto-cq committed:
UIControlの見た目を修正

・テキストフィールドが見えにくいので枠に線と色を指定
・ボタン、テキストフィールドを角丸に修正

khatamoto-cq committed:
ログイン後、一覧画面がモーダルでページが表示されるのを修正

ykawakami-cq commented on BookManager/EditBookViewController.swift:
```diff
@@ -0,0 +1,83 @@
+    @IBAction func attachImageAction(_ sender: Any) {
+        let photo = UIImagePickerControllerSourceType.photoLibrary
+
+        if UIImagePickerController.isSourceTypeAvailable(photo) {
```
こちらも同じく省略できます。

ykawakami-cq replied:
こちらも同じく省略できます。

ykawakami-cq commented on BookManager/AddBookViewController.swift:
```diff
@@ -0,0 +1,86 @@
+    @IBAction func attachImageAction(_ sender: Any) {
+        let photo = UIImagePickerControllerSourceType.photoLibrary
+
+        if UIImagePickerController.isSourceTypeAvailable(photo) {
```
UIImagePickerControllerSourceType省略して下記のように書けます。
```
if UIImagePickerController.isSourceTypeAvailable(.photoLibrary) {
}
```

ykawakami-cq commented on BookManager/AddBookViewController.swift:
```diff
@@ -0,0 +1,86 @@
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyy/MM/dd"
+        purchaseDateTextField.text = dateFormatter.string(from: sender.date)
+    }
```
`AddBookViewController`と`EditBookViewController`で同じメソッドを使っているので、
ユーティリティクラスを作ってまとめたほうが良いかなと思います。

ykawakami-cq commented on BookManager/AppDelegate.swift:
```diff
@@ -13,34 +13,44 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     var window: UIWindow?
 
+    func application(_ application: UIApplication,
+                     didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
```
動作確認の際に見落としてしまったかもしれないのですが、初回起動時にアカウント設定画面を表示する処理ないですよね？

ykawakami-cq replied:
動作確認の際に見落としてしまったかもしれないのですが、初回起動時のみ「アカウント設定」を表示するようにお願いします。

ykawakami-cq commented on BookManager/ListBooksViewController.swift:
```diff
@@ -0,0 +1,49 @@
+    var books = Book.allBooks
+
+    @IBAction func tapAddAction(_ sender: Any) {
+        performSegue(withIdentifier: "addBook", sender: self)
```
Segueの名前が変更になった時に、直し忘れていてもコンパイル時にチェックできない等、
危うい点もあるのでコードで遷移するようにお願いします。

ykawakami-cq reviewed (commented):
確認お願いします。

khatamoto-cq committed:
Segueを使わずに遷移させるよう修正

khatamoto-cq committed:
ファイルの添付とDatePickerを共通化した

khatamoto-cq committed:
Xcodeが生成したコメントを削除

khatamoto-cq committed:
初回起動時のみ「アカウント設定」を表示するように修正

ykawakami-cq commented on BookManager/EditBookViewController.swift:
```diff
@@ -0,0 +1,64 @@
+    func pickerChanged(sender: UIDatePicker) {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyy/MM/dd"
+        purchaseDateTextField.text = dateFormatter.string(from: sender.date)
```
日付のフォーマットも`DatePickerHelper`にまとめてしまっても良いかもしれません

khatamoto-cq committed:
DatePickerとUIImagePickerのコールバックも共通化

khatamoto-cq committed:
モデルのプロパティを正しく修正

khatamoto-cq committed:
一覧画面から編集画面に値を渡すよう修正

ykawakami-cq commented on BookManager/ListBooksViewController.swift:
```diff
@@ -0,0 +1,52 @@
+        super.didReceiveMemoryWarning()
+    }
+
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
```
ここもSegueではなく`didSelectRowAt`を使うようにお願いします。

khatamoto-cq committed:
書籍一覧から書籍編集の画面遷移をsegueを使わない実装に変更

khatamoto-cq committed:
設定画面のsegueの遷移をやめた

ykawakami-cq commented on BookManager/AccountViewController.swift:
```diff
@@ -0,0 +1,63 @@
+
+    func saveLoginState() {
+        let userDefaults = UserDefaults.standard
+        userDefaults.set(true, forKey: "logined")
```
下記のように1行でできるかなと思います。
`UserDefaults.standard.set(true, forKey: "logined")`

khatamoto-cq replied:
シングルトンだったんですね。sharedみたいなメソッドではないので一旦変数に格納してました。
ドキュメントみたら、「Returns the shared defaults object」ってちゃんと書いてありました。ありがとうございます。

ykawakami-cq commented on BookManager/AddBookViewController.swift:
```diff
@@ -0,0 +1,63 @@
+    @IBAction func popupDatepickerAction(_ sender: UITextField) {
+        DatePickerHelper.load(self, action: #selector(pickerChanged), sender: sender)
+
+    }
```
余分な改行が入ってますね

khatamoto-cq committed:
framework(APIKit, Himotoki, Kingfisher)の導入

・Cartfileを追加
・Cartfile/以下は管理しないので.gitignoreに追記
・SwiftLintがframework内の警告を出すようになったので.swiftlint.ymlにて対象外にした

ykawakami-cq commented on BookManager.xcodeproj/project.pbxproj:
```diff
@@ -7,21 +7,41 @@
 	objects = {
 
 /* Begin PBXBuildFile section */
+		231B23901EE6BEDA004C3221 /* AddBookViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 231B238F1EE6BEDA004C3221 /* AddBookViewController.swift */; };
```
テスト

khatamoto-cq committed:
レビュー指摘箇所を修正

・余分な改行を削除
・UserDefaultsの余分な定数を削除

khatamoto-cq committed:
Merge branch 'ui' into api

khatamoto-cq committed:
SwiftLintが原因でR.swiftがインストールできない問題を解決

・.swiftlint.ymlでR.generated.swiftを対象外にした
・rswiftのバージョンを最新にした

ykawakami-cq commented:
@khatamoto-cq 
問題無さそうです。

khatamoto-cq committed:
R.swiftを使うように修正

hyoshimi-cq commented on BookManager/AccountViewController.swift:
```diff
@@ -0,0 +1,62 @@
+            closeButton.isHidden = false
+        } else {
+            closeButton.isEnabled = false
+            closeButton.isHidden = true
```
```swift
let enabled = UserDefaults.standard.bool(forKey: "logined")
closeButton.isEnabled = enabled
closeButton.isHidden = !enabled
```
上記のようにすればif文を使わなくていいので少し見やすくなると思います。変数名の `enabled` が少し微妙かもしれませんが、、、

khatamoto-cq replied:
確かにコードが少なくなりますね。ご指摘ありがとうございます。

hyoshimi-cq commented on BookManager/AccountViewController.swift:
```diff
@@ -0,0 +1,62 @@
+        passwordConfirmTextField.layer.borderWidth = 1
+        passwordConfirmTextField.layer.cornerRadius = 5
+        passwordConfirmTextField.layer.borderColor = UIColor.lightGray.cgColor
+        passwordConfirmTextField.layer.masksToBounds = true
```
メール、パスワード、パスワード再入力の設定を全て同じにするのであれば定数にして意味を持たせた方がいいかもしれないです。

hyoshimi-cq commented on BookManager/Utility/FileAttachHelper.swift:
```diff
@@ -0,0 +1,22 @@
+            imageView.image = image
+        }
+
+        controller.dismiss(animated: true)
```
写真ライブラリから画像を選択した後に書籍追加・編集画面が閉じられちゃいます？

khatamoto-cq replied:
いえ、テストしてますが閉じられません。controller引数にはUIViewControllerの参照を渡しておりますが、この行がなければCamera Rollが閉じなくなります。

hyoshimi-cq commented on BookManager/LoginViewController.swift:
```diff
@@ -0,0 +1,53 @@
+        if validate() {
+            let controller = R.storyboard.main.tabViewController()
+            controller?.modalTransitionStyle = .crossDissolve
+            present(controller!, animated: true, completion: nil)
```
細かい話なのでどちらでもいいのですが、ここで早期returnしておくとelse節がなくなって見やすくなります。

khatamoto-cq replied:
ネストが深くならなくなりました。ご指摘ありがとうございます。

hyoshimi-cq commented on BookManager/Book.swift:
```diff
@@ -0,0 +1,44 @@
+        self.name = dictionay[Book.Name]!
+        self.price = dictionay[Book.Price]!
+        self.purchaseDate = dictionay[Book.PurchaseDate]!
+        self.imageUrl = dictionay[Book.ImageUrl]!
```
`dictionay` に足りない要素があると困るので4つの値を引数に持つ方がいいのかなと思いました。

hyoshimi-cq reviewed (commented):
レビューしました。
被っている箇所はコメントしてません。
よろしくお願いします。

khatamoto-cq committed:
レビュー指摘箇所を修正

khatamoto-cq commented:
@h.yoshimi

レビューありがとうございました。
ご指摘頂いた箇所を修正いたしました。ご確認頂けますでしょうか。

hyoshimi-cq commented:
@khatamoto-cq 
問題ないと思います！

khatamoto-cq commented:
@hyoshimi-cq 
ありがとうございます。

miyoshitakaaki commented on BookManager/AccountViewController.swift:
```diff
@@ -0,0 +1,58 @@
+    @IBAction func saveAction(_ sender: Any) {
+        print("アカウント設定を登録しました")
+
+        if UserDefaults.standard.bool(forKey: "logined") == false {
```
```
UserDefaults.standard.bool(forKey: "logined") == false
```
は
```
!UserDefaults.standard.bool(forKey: "logined")
```
こう書けます

miyoshitakaaki commented on BookManager/AccountViewController.swift:
```diff
@@ -0,0 +1,58 @@
+        emailTextField.layer.borderWidth = Const.TextFieldBorderWidth
+        emailTextField.layer.cornerRadius = Const.TextFieldCornerRadius
+        emailTextField.layer.borderColor = Const.TextFieldBorderColor
+        emailTextField.layer.masksToBounds = Const.TextFieldMasksToBounds
```
この辺り似たような処理があるので、なんとかしたいですね
一つ方法としてはUITextFieldを拡張してやるといいと思います。
```
extension UITextField {
  fun hoge(width : CGFloat, radius: CGFloat) {
    self.layer. borderWidth = width
    self.layer. cornerRadius = radius
    // 省略
  }
}
```

miyoshitakaaki replied:
こうしてやると、
```
emailTextField.hoge(Const.TextFieldBorderWidth, Const.TextFieldCornerRadius)
```
な感じで呼べます

khatamoto-cq replied:
すごいコード量が減りましたし、読みやすくなりました。
ご指摘ありがとうございます。

miyoshitakaaki commented on BookManager/AccountViewController.swift:
```diff
@@ -0,0 +1,58 @@
+
+class AccountViewController: UIViewController {
+
+    @IBOutlet weak var emailTextField: UITextField!
```
好みの問題ですが、
```
@IBOutlet weak var emailTextField: UITextField! {
  didSet {
        emailTextField.layer.borderWidth = Const.TextFieldBorderWidth
        emailTextField.layer.cornerRadius = Const.TextFieldCornerRadius
        emailTextField.layer.borderColor = Const.TextFieldBorderColor
        emailTextField.layer.masksToBounds = Const.TextFieldMasksToBounds
  }
}
```
な感じにしてやると、プロパティ宣言と属性指定を一箇所に固められます。
他も同様にしてやると、initControlsメソッド無くせますね

khatamoto-cq replied:
IBOutletとそのアトリビュートが一緒になっているのでコードが分かりやすくなりました。
ご指摘ありがとうございます。

miyoshitakaaki commented on BookManager/AddBookViewController.swift:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @IBAction func attachImageAction(_ sender: Any) {
+        FileAttachHelper.load(self, delegate: self)
```
selfを第一引数と第二引数で２つ渡しているのは不自然です、１つだけで

khatamoto-cq replied:
メソッド中でキャストすれば良いということですね。
ご指摘ありがとうございます。

miyoshitakaaki commented on BookManager/AddBookViewController.swift:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
+        FileAttachHelper.setImage(self, imageView: self.imageView, info: info)
```
上記と同じ理由で、selfを渡しているならself.imageはsetImageメソッド内で取得できるので、
imageView引数は不要です

khatamoto-cq replied:
AddBookViewControllerとEditBookViewControllerの両方から使われ、引数の型をUIViewControllerにしているのですが、selfがメソッド内に入るとどちらのViewControllerからのものか分からなくなってしまいます。
解決する良い方法があればご教示頂けませんでしょうか。

miyoshitakaaki replied:
setImageメソッド内でimageViewを取得したいということですかね？
であれば、最善かはわからないですがプロトコルを作ることで解決できます
```
protocol FileAttachable {
  val imageView: UIImageView { get }
}
```
的なのを作っておいて、ViewControllerに準拠させて
setImageの引数をFileAttachableにするという方針
↑は試してないので、修正が必要かもです

miyoshitakaaki commented on BookManager/Book.swift:
```diff
@@ -0,0 +1,45 @@
+    static let PurchaseDate = "purchaseDate"
+    static let ImageUrl = "imageUrl"
+
+    init(name: String, price: String, purchaseDate: String, imageUrl: String) {
```
structはclassと違って、イニシャライザを定義しなくてもいい特徴があります
このイニシャライザは不要です

miyoshitakaaki commented on BookManager/Book.swift:
```diff
@@ -0,0 +1,45 @@
+        return books
+    }
+
+    static func load() -> [[String : String]] {
```
[[String : String]]
のような多次元配列（中身は辞書型ですが）は避けた方がいいです
今は改善されたかもしれませんが、ビルドが不安定になります

khatamoto-cq replied:
ダミーデータを作る箇所でしたので、実装をあまり気にしていませんでしたが、ご指摘されて修正してみるとだいぶコード量が減って分かりやすくなりました。ありがとうございます。

miyoshitakaaki commented on BookManager/LoginViewController.swift:
```diff
@@ -0,0 +1,53 @@
+                        message: "メールアドレスもしくはパスワードが違います", preferredStyle: .alert)
+        let canselAction = UIAlertAction(title: "OK", style: .cancel, handler: nil)
+        alertWindow.addAction(canselAction)
+        present(alertWindow, animated: true, completion: nil)
```
前方が正常系で、後方は非正常系は違和感が。。。
guardで非正常系を早期リターンで、最後に正常系で 🙏

miyoshitakaaki commented on BookManager/TabViewController.swift:
```diff
@@ -0,0 +1,17 @@
+import UIKit
+
+class TabViewController: UITabBarController, UITabBarControllerDelegate {
```
ここに全てのプロトコルを書くと、クラス内にデリゲートメソッドが大量に増えてくるので、
TabViewControllerを拡張して、そこにUITabBarControllerDelegateメソッドを固めてやりましょう

```
func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController)
```
はUITabBarControllerDelegateのデリゲートメソッド？だった気がするので、
```
extension TabViewController: UITabBarControllerDelegate {
  // UITabBarControllerDelegateのデリゲートメソッド
  func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController)
}
```

khatamoto-cq replied:
extensionはこういうこともできるんですね。TableViewはUITableViewControllerを継承して使っているんですが、もしUITableViewDelegate, UITableViewDataSourceのプロトコルの実装をUIViewControllerにするような場合は、この方法を使ってViewControllerから処理を分けたらよいのでしょうか？

miyoshitakaaki replied:
> もしUITableViewDelegate, UITableViewDataSourceのプロトコルの実装をUIViewControllerにするような場合は、この方法を使ってViewControllerから処理を分けたらよいのでしょうか？

そうですね、分けた方がわかりやすいです。
デリゲートメソッドが一見どのprotocolに属しているのかがわからない時があるので。。。

miyoshitakaaki commented on BookManager/Utility/Const.swift:
```diff
@@ -0,0 +1,17 @@
+//
+//  Created by hatamoto kanji on 2017/06/12.
+//  Copyright © 2017年 hatamoto kanji. All rights reserved.
+//
```
署名は消して下さい 🙇

miyoshitakaaki commented:
@khatamoto-cq 
一旦レビューしました、確認お願いします 🙏

khatamoto-cq committed:
署名の消し忘れを削除

khatamoto-cq committed:
IBOutletのアトリビュートをdidSetに纏めた

khatamoto-cq committed:
「式 == false」の書き方を「!式」に変更

khatamoto-cq committed:
UIControlのextentionを作成してコードを少なくした

khatamoto-cq committed:
Lintエラーを修正

khatamoto-cq committed:
不要なメソッドの引数を削除

khatamoto-cq committed:
ビルドエラーを修正

khatamoto-cq committed:
不要なStructのイニシャライザを削除

khatamoto-cq committed:
多次元配列を廃止

khatamoto-cq committed:
異常系を先にチェックして、その後正常系を記述するよう修正

khatamoto-cq committed:
extensionを使ってデリゲートのプロトコルを実装するように修正

khatamoto-cq commented:
@miyoshi-cq 
レビューありがとうございました。
selfでImageViewを取得する箇所以外は修正いたしました。保留した件はもう少し調べてみます。

miyoshitakaaki commented on BookManager/AccountViewController.swift:
```diff
@@ -0,0 +1,56 @@
+        }
+    }
+
+    @IBOutlet weak var closeButton: UIButton!
```
ここもついでに
```
@IBOutlet weak var closeButton: UIButton! {
  didSet {
        let logined = UserDefaults.standard.bool(forKey: "logined")
        closeButton.isEnabled = logined
        closeButton.isHidden = !logined
  }
}
```
にできないですかね？

miyoshitakaaki commented on BookManager/EditBookViewController.swift:
```diff
@@ -0,0 +1,66 @@
+import UIKit
+
+class EditBookViewController: UIViewController, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
```
ここも
```
extension EditBookViewController : UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
        FileAttachHelper.setImage(self, imageView: self.imageView, info: info)
     }  
}
```
としたいところです

miyoshitakaaki replied:
UIImagePickerControllerDelegateとUINavigationControllerDelegate
はワンセットにしないといけなかった気がするのでこうしていますが、
基本はextension一つに対して、protocol一つ指定で 🙇

miyoshitakaaki commented on BookManager/ListBooksViewController.swift:
```diff
@@ -0,0 +1,47 @@
+        cell.bookImageView.image = UIImage(named: book.imageUrl)
+        cell.nameLabel.text = book.name
+        cell.priceLabel.text = book.price
+        cell.purchaseDateLabel.text = book.purchaseDate
```
cellにsetData的なメソッドをはやして、Bookを引数にとれればすっきりしそうです

miyoshitakaaki commented on BookManager/Book.swift:
```diff
@@ -0,0 +1,35 @@
+    static let Name = "name"
+    static let Price = "price"
+    static let PurchaseDate = "purchaseDate"
+    static let ImageUrl = "imageUrl"
```
この辺りってまだ必要ですかね？
いらない気がするのですがいかがでしょうか

miyoshitakaaki commented on BookManager/LoginViewController.swift:
```diff
@@ -0,0 +1,51 @@
+
+    @IBAction func tapLoginAction(_ sender: Any) {
+        guard validate() else {
+            let alertWindow = UIAlertController(title: "ログインエラー",
```
文字列はLocalizable.stringを使って一元管理してほしいです cc: @hyoshimi-cq

miyoshitakaaki commented on BookManager/LoginViewController.swift:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    func validate() -> Bool {
+        if emailTextField.text == "" || passwordTextField.text == "" {
```
空文字判定は
```
emailTextField.text.isEmpty
```
でお願いします

miyoshitakaaki commented on BookManager/Utility/DatePickerHelper.swift:
```diff
@@ -0,0 +1,18 @@
+class DatePickerHelper {
+    class func load(_ target: Any?, action: Selector, sender: UITextField) {
+        let datePicker = UIDatePicker()
+        datePicker.datePickerMode = UIDatePickerMode.date
```
UIDatePickerModeは省略できます

miyoshitakaaki commented on BookManager/AddBookViewController.swift:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    @IBAction func closeAction(_ sender: Any) {
```
IBActionのメソッド名の末尾にActionは不要かなって思います
iOSはwillやdidを接頭辞として用いるのが慣習なので、
didCloseButtonTappedやwillCloseScreenなどがよいかなと思います
個人的にはIBActionはdid付きのほうが好きですが

khatamoto-cq committed:
アカウント設定の閉じるボタンのisEnable, isHiddenの制御もdidSetにまとめた

khatamoto-cq committed:
UIImagePickerControllerDelegate, UINavigationControllerDelegateの実装をextensionにまとめた

khatamoto-cq committed:
TableViewCellにモデルを渡してセル内のアトリビュートへの代入はTableCell内でやるように修正

khatamoto-cq committed:
Delegate内の条件式を良いイディオムを使うように修正

khatamoto-cq committed:
以前使っていたdictionayのキーの消し忘れを削除

khatamoto-cq committed:
空文字チェックはString#isEmptyを使うように修正

khatamoto-cq committed:
Enumは省略形に修正

khatamoto-cq committed:
@IBActionの命名規則を変更

khatamoto-cq committed:
setImageにImageViewを渡さないように実装

khatamoto-cq committed:
Localizable.stringsに対応

khatamoto-cq committed:
LocalizedString化の漏れを追加

khatamoto-cq commented:
@miyoshi-cq 
ご指摘頂いた箇所はを修正いたしました。
ご確認頂けますでしょうか。

miyoshitakaaki commented on BookManager/LoginViewController.swift:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    func validate() -> Bool {
+        if (emailTextField.text?.isEmpty)! || (passwordTextField.text?.isEmpty)! {
```
```
(emailTextField.text?.isEmpty)!
```
って書くと、emailTextField.textがnilの時クラッシュしますので、
API周り対応する際ご注意いただければと

miyoshitakaaki commented on BookManager/LoginViewController.swift:
```diff
@@ -0,0 +1,51 @@
+
+    @IBAction func didLoginButtonTapped(_ sender: Any) {
+        guard validate() else {
+            let alertWindow = UIAlertController(title: NSLocalizedString("LoginErrorTitle", comment: ""),
```
R.swiftはNSLocalizedStringをサポートしていて
```
R.string.localizable.LoginErrorTitle()
```
と書けます
タイポを防げます

khatamoto-cq replied:
NSLocalizedStringの「comment: ""」の部分がなんとかならないかと思っておりました。これならタイポも防げるだけでなくきれいです。
ご指摘ありがとうございます。

miyoshitakaaki commented on BookManager/Utility/UIControlExtention.swift:
```diff
@@ -0,0 +1,17 @@
+import UIKit
+
+extension UITextField {
+    func ext(borderWidth: CGFloat, borderColor: CGColor, radius: CGFloat, masksToBound: Bool) {
```
extっていうメソッドちょっとわからないので、
属性指定という意味で、applyとかですかね

miyoshitakaaki commented:
@khatamoto-cq 
大体問題ないと思うので、修正したらマージしちゃって大丈夫です 👍

khatamoto-cq committed:
NSLocalizedStringをR.swfitで書き換え

khatamoto-cq committed:
TextFieldのextensionをextからapplyに変更

khatamoto-cq commented:
@miyoshi-cq 

修正いたしました。
レビューありがとうございました。masterにマージ致します。
